contracts/VestingManager.sol

```diff
@@ -0,0 +1,212 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.0;
+
+import '@openzeppelin/contracts/access/Ownable.sol';
+import '@openzeppelin/contracts/token/ERC20/ERC20.sol';
+import '@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol';
+
+import 'hardhat/console.sol';
+
+contract VestingManager is Ownable {
+    using SafeERC20 for IERC20;
+
+    IERC20 public immutable oidToken;
+    uint256 public referenceDate = 0;
+    uint32 public constant QUARTER = 13 weeks;
+    uint8 public schemaId = 0;
+
+    struct VestingSchema {
+        uint8 lockup; // Number of quarters during which the holders tokens are locked and not accountable for vesting
+        uint16 vesting; // % of total tokens released every quarter with two decimals of resolution (i.e. 1256 -> 12,56%)
+    }
+
+    struct Holding {
+        uint256 totalVestedAmount;
+        uint256 releasedAmount;
+        uint8 vestingSchema;
+    }
+
+    /* Mappings */
+    mapping(uint8 => VestingSchema) public VestingSchemas;
+    mapping(address => Holding[]) public Holdings;
+
+    /* Events */
+    event Deposit(address to, uint256 amount, uint8 schemaId);
+    event Withdraw(address to, uint256 amount);
+    event SchemaCreated(uint8 schemaId, uint8 lockup, uint16 vesting);
+    event EthDeposit(address sender, uint256 value, uint256 balance);
+
+    /* Modifiers */
+    modifier referenceInitiated() {
+        require(referenceDate != 0, 'Reference date not initialized');
+        _;
+    }
+
+    constructor(IERC20 oidToken_) {
+        require(address(oidToken_) != address(0), 'Token address cannot be addr(0)');
+        oidToken = oidToken_;
+    }
+
+    /* Admin functions */
+
+    /**
+     * @dev Sets the reference date for all lockups in the contract.
+     * @param _referenceDate period in quarters that tokens remain blocked
+     */
+    function setReferenceDate(uint256 _referenceDate) external onlyOwner {
+        require(referenceDate == 0, 'Reference date has already been initialized');
+        require(_referenceDate != 0, 'Cannot set reference date to 0');
+        referenceDate = _referenceDate;
+    }
+
+    /**
+     * @dev Default withdraw function in case someone deposits ETH erroneously.
+     */
+    function withdraw() external onlyOwner {
```

Comment:
not sure this is needed

Reply:
Implemented to prevent ETH permanently stuck in the contract whatever the cause